File: backend/src/models/questions/questions.py
from pydantic import Field, BaseModel
from typing import List, Dict, Union, Literal, Optional
from models._base import MongoBaseModel

class BaseQuestion(MongoBaseModel):
    """Base question model that all question types inherit from"""
    type: str = Field(..., description="Question type")

class MultipleChoiceQuestion(BaseQuestion):
    type: Literal["multiple_choice"] = Field(default="multiple_choice", description="Question type")
    question: str = Field(..., description="The question text")
    options: List[str] = Field(..., description="List of answer options")
    correctOption: int = Field(..., description="Index of the correct option")
    explanation: str = Field(..., description="Explanation for the correct answer")

class TrueFalseQuestion(BaseQuestion):
    type: Literal["true_false"] = Field(default="true_false", description="Question type")
    statement: str = Field(..., description="The statement to evaluate")
    answer: bool = Field(..., description="Whether the statement is true or false")
    explanation: str = Field(..., description="Explanation for the answer")

class FillBlankQuestion(BaseQuestion):
    type: Literal["fill_blank"] = Field(default="fill_blank", description="Question type")
    question: str = Field(..., description="The question text with blank")
    answer: str = Field(..., description="The correct answer to fill the blank")
    validation: Literal["exact", "case_insensitive", "contains"] = Field(
        default="exact", 
        description="Type of validation for the answer"
    )
    explanation: str = Field(..., description="Explanation for the correct answer")

class OrderQuestion(BaseQuestion):
    type: Literal["order"] = Field(default="order", description="Question type")
    question: str = Field(..., description="The question text")
    items: List[str] = Field(..., description="List of items to be ordered")
    correctOrder: List[int] = Field(..., description="Correct order as list of indices")
    explanation: Optional[str] = Field(default="", description="Explanation for the correct order")

class MatchQuestion(BaseQuestion):
    type: Literal["match"] = Field(default="match", description="Question type")
    question: str = Field(..., description="The question text")
    pairs: Dict[str, str] = Field(..., description="Dictionary of items to match")
    explanation: Optional[str] = Field(default="", description="Explanation for the matches")

# Union type for all question types
Question = Union[
    MultipleChoiceQuestion,
    TrueFalseQuestion,
    FillBlankQuestion,
    OrderQuestion,
    MatchQuestion
]

class QuestionCreate(BaseModel):
    type: Literal["multiple_choice", "true_false", "fill_blank", "order", "match"]
    question: Optional[str] = None
    statement: Optional[str] = None  # For true_false
    options: Optional[List[str]] = None  # For multiple_choice
    correctOption: Optional[int] = None  # For multiple_choice
    answer: Optional[Union[str, bool]] = None  # For fill_blank and true_false
    validation: Optional[str] = None  # For fill_blank
    items: Optional[List[str]] = None  # For order
    correctOrder: Optional[List[int]] = None  # For order
    pairs: Optional[Dict[str, str]] = None  # For match
    explanation: Optional[str] = None

class QuestionListCreate(BaseModel):
    questions: List[QuestionCreate] = Field(..., description="List of questions to create")

File: frontend/src/_interfaces/questions/questions.ts
export interface BaseQuestion {
    id: string;
    type: string;
    created_at?: string;
    updated_at?: string;
}

export interface MultipleChoiceQuestion extends BaseQuestion {
    type: "multiple_choice";
    question: string;
    options: string[];
    correctOption: number;
    explanation: string;
}

export interface TrueFalseQuestion extends BaseQuestion {
    type: "true_false";
    statement: string;
    answer: boolean;
    explanation: string;
}

export interface FillBlankQuestion extends BaseQuestion {
    type: "fill_blank";
    question: string;
    answer: string;
    validation: "exact" | "case_insensitive" | "contains";
    explanation: string;
}

export interface OrderQuestion extends BaseQuestion {
    type: "order";
    question: string;
    items: string[];
    correctOrder: number[];
    explanation?: string;
}

export interface MatchQuestion extends BaseQuestion {
    type: "match";
    question: string;
    pairs: Record<string, string>;
    explanation?: string;
}

export type Question = 
    | MultipleChoiceQuestion 
    | TrueFalseQuestion 
    | FillBlankQuestion 
    | OrderQuestion 
    | MatchQuestion;

export interface QuestionsResponse {
    questions?: Question[];
    message?: string;
}

export interface QuestionParams {
    skip?: number;
    limit?: number;
}

export type QuestionType = "multiple_choice" | "true_false" | "fill_blank" | "order" | "match";